# import Library

In [1]:
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import Dataset
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch.utils.tensorboard import SummaryWriter

from PIL import Image
from collections import OrderedDict #for multiGPU
from tqdm import tqdm

import os
import random
import time

# GPU

In [2]:
# device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
#for colab
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print("device is ", device)

device is  cuda


# Load Data

In [3]:
print("loading dataset")

data_transforms = {
    'train' : transforms.Compose([ #horizontal flip, random crop, normalization           
            #transforms.ToPILImage(),
            # transforms.Resize((256,256)),
            transforms.RandomCrop(32, padding = 4), #augmentation, input for ResNet
            transforms.RandomHorizontalFlip(), #augmentation
            transforms.ToTensor(),
            transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
            #transforms.Normalize((0.4914, 0.4822, 0.4465), (1, 1, 1)), #with the per-pixel mean subtracted

        ]) , 
        'test': transforms.Compose([
            transforms.ToTensor(), #single view of the original 32x32 image
            transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
        ])
    }
#access data_transforms['train']

#dataloader (batch)
train_set = torchvision.datasets.CIFAR10(
    root='./data', train=True, download=True, transform = data_transforms['train']
)
trainloader = torch.utils.data.DataLoader(
    train_set, batch_size = 128, shuffle = True, num_workers=2
)

test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)

#classes
classes = ('plane', 'car', 'bird', 'cat', 'deer','dog', 'frog', 'horse', 'ship', 'truck')


loading dataset


100%|██████████| 170498071/170498071 [00:01<00:00, 102857450.01it/s]


Extracting ./data/cifar-10-python.tar.gz to ./data
Files already downloaded and verified


# Model Architecture

In [4]:
class optionA(nn.Module):
  def __init__(self, func):
    super(optionA, self).__init__()
    self.func = func
  
  def forward(self, x):
    return self.func(x)

In [5]:
class Fx2_16(nn.Module):
  def __init__(self):
    super().__init__()
    self.conv1 = nn.Sequential(
        nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(16)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(16)
    )
    self.residual = nn.Sequential()
  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out +=self.residual(x)
    out = F.relu(out)

    return out

In [6]:
class Fx2_32(nn.Module):
  def __init__(self):
    super().__init__()
    self.conv1 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32),
    )
    self.residual = nn.Sequential()
  
  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out += self.residual(x)
    out = F.relu(out)
    return out


In [7]:
class Fx2_64(nn.Module):
  def __init__(self):
    super().__init__()

    self.conv1 = nn.Sequential(
         nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.residual = nn.Sequential()

  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out += self.residual(x)
    out = F.relu(out)
    return out

In [8]:
#classes for Resnet56 Model
class conv16 (nn.Module): #Blue
  def __init__(self):
    super().__init__()
    self.Fx1 = Fx2_16()
    self.Fx2 = Fx2_16()
    self.Fx3 = Fx2_16()
    self.Fx4 = Fx2_16()
    self.Fx5 = Fx2_16()
    self.Fx6 = Fx2_16()
    self.Fx7 = Fx2_16()
    self.Fx8 = Fx2_16()
    self.Fx9 = Fx2_16()
    
  def forward(self,x):
    x1 = self.Fx1(x)
    x2 = self.Fx2(x1)
    x3 = self.Fx3(x2) 
    x4 = self.Fx4(x3) 
    x5 = self.Fx5(x4) 
    x6 = self.Fx6(x5) 
    x7 = self.Fx7(x6) 
    x8 = self.Fx8(x7) 
    x9 = self.Fx9(x8)
    #x = self.relu(x)
    return x9

#classes for Resnet56 Model
class conv32 (nn.Module): #Purple
  def __init__(self):
    super().__init__()
    self.Fx1_1 = nn.Sequential(
        #torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', device=None, dtype=None)
        nn.Conv2d(16, 32, kernel_size = 3, stride = 2, padding = 1, bias = False),
        nn.BatchNorm2d(32))
    self.Fx1_2 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32)
    )

    ## OptionB
    # self.residual_mapping = nn.Sequential(
    #     nn.Conv2d(16,32,kernel_size =1, stride = 2, bias= False),
    #     nn.BatchNorm2d(32)
    # )

    # ## OptionA
    self.residual_mapping = optionA(lambda x: F.pad(x[:, :, ::2, ::2], (0, 0, 0, 0, 32//4, 32//4), "constant", 0))

    self.Fx2 = Fx2_32()
    self.Fx3 = Fx2_32()
    self.Fx4 = Fx2_32()
    self.Fx5 = Fx2_32()
    self.Fx6 = Fx2_32()
    self.Fx7 = Fx2_32()
    self.Fx8 = Fx2_32()
    self.Fx9 = Fx2_32()
    # self.relu = nn.ReLU()
    
  def forward(self,x):
    out = F.relu(self.Fx1_1(x))
    out = self.Fx1_2(out) 
    out += self.residual_mapping(x)
    x1 = F.relu(out)
    
    x2 = self.Fx2(x1) 
    x3 = self.Fx3(x2) 
    x4 = self.Fx4(x3) 
    x5 = self.Fx5(x4) 
    x6 = self.Fx6(x5) 
    x7 = self.Fx7(x6) 
    x8 = self.Fx8(x7) 
    x9 = self.Fx9(x8) 
    # x2 = self.Fx2(x1) + x1
    # x3 = self.Fx3(x2) + x2
    #x = self.relu(x)
    return x9

#classes for Resnet56 Model
class conv64 (nn.Module): #Green
  def __init__(self):
    super().__init__()
    self.Fx1_1 = nn.Sequential(
        #torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', device=None, dtype=None)
        nn.Conv2d(32, 64, kernel_size = 3, stride = 2, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.Fx1_2 = nn.Sequential(
        nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )

    ## OptionB
    # self.residual_mapping = nn.Sequential(
    #     nn.Conv2d(32,64,kernel_size =1, stride = 2, bias= False),
    #     nn.BatchNorm2d(64)
    # )

    ## OptionA
    self.residual_mapping = optionA(lambda x: F.pad(x[:, :, ::2, ::2], (0, 0, 0, 0, 64//4, 64//4), "constant", 0))
    
    self.Fx2 = Fx2_64()
    self.Fx3 = Fx2_64()
    self.Fx4 = Fx2_64()
    self.Fx5 = Fx2_64()
    self.Fx6 = Fx2_64()
    self.Fx7 = Fx2_64()
    self.Fx8 = Fx2_64()
    self.Fx9 = Fx2_64()

    # self.relu = nn.ReLU()
    
  def forward(self,x):
    out = F.relu(self.Fx1_1(x))
    out = self.Fx1_2(out)
    out += self.residual_mapping(x)
    x1 = F.relu(out)
    x2 = self.Fx2(x1) 
    x3 = self.Fx3(x2) 
    x4 = self.Fx4(x3) 
    x5 = self.Fx5(x4)
    x6 = self.Fx6(x5) 
    x7 = self.Fx7(x6) 
    x8 = self.Fx8(x7) 
    x9 = self.Fx9(x8) 

    return x9
   


In [9]:
class MyModel(nn.Module): #Resnet56 architecture implementation
    def __init__(self, conv2 = conv16, conv3 = conv32,
            conv4 = conv64, num_classes = 10, init_weights = True): #num_classes
        super().__init__()
        self.conv1 = nn.Sequential(
            nn.Conv2d(3, 16, kernel_size = 3, stride = 1, padding=1, bias = False),
            nn.BatchNorm2d(16)
        )
        self.conv2 = conv2()
        self.conv3 = conv3()
        self.conv4 = conv4()
        #self.avg_pool = nn.AdaptiveAvgPool2d((1,1))
        self.fclayer = nn.Linear(64, num_classes)

        if init_weights:
            self.initialize_weights()
    
    # define weight initialization function
    def initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, 0, 0.01)
                nn.init.constant_(m.bias, 0) 
        
    def forward(self, x):
        x1 = F.relu(self.conv1(x))
        x2 = self.conv2(x1)
        x3 = self.conv3(x2)
        x4 = self.conv4(x3)
        x = F.avg_pool2d(x4, x4.size()[3])
        #x = self.avg_pool(x4)
        x = x.view(x.size(0), -1)
        output = self.fclayer(x)
        return output

In [ ]:
resnet56 = MyModel()

In [ ]:
from torchsummary import summary
summary(resnet56, (3,32,32))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 32, 32]             432
       BatchNorm2d-2           [-1, 16, 32, 32]              32
            Conv2d-3           [-1, 16, 32, 32]           2,304
       BatchNorm2d-4           [-1, 16, 32, 32]              32
            Conv2d-5           [-1, 16, 32, 32]           2,304
       BatchNorm2d-6           [-1, 16, 32, 32]              32
            Fx2_16-7           [-1, 16, 32, 32]               0
            Conv2d-8           [-1, 16, 32, 32]           2,304
       BatchNorm2d-9           [-1, 16, 32, 32]              32
           Conv2d-10           [-1, 16, 32, 32]           2,304
      BatchNorm2d-11           [-1, 16, 32, 32]              32
           Fx2_16-12           [-1, 16, 32, 32]               0
           Conv2d-13           [-1, 16, 32, 32]           2,304
      BatchNorm2d-14           [-1, 16,

In [ ]:
MyModel()

MyModel(
  (conv1): Sequential(
    (0): Conv2d(3, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
  (conv2): conv16(
    (Fx1): Fx2_16(
      (conv1): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (residual): Sequential()
    )
    (Fx2): Fx2_16(
      (conv1): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Sequential(
        (0)

# Train

In [10]:
# #get_model when loading model parameters trained w/ multi-gpu
def get_model(model_name, checkpoint_path):
    print("start loading model parameter")
    model = model_name()
    state_dict = torch.load(checkpoint_path)

    new_state_dict = OrderedDict()

    for k, v in state_dict.items():
        name = k.replace('module.', '')
        new_state_dict[name] = v
    model.load_state_dict(new_state_dict)
    print("model parameter loaded")
    return model

In [11]:

def train(n_epochs, train_loader) :
 
    epochs = n_epochs
    train_dataloader = train_loader
    #get current time -> folder name
    train_date = time.strftime("%y%m%d_%H%M", time.localtime(time.time()))
    # print(train_date)

    #log path for tensorboard
    log_path = os.path.join(os.getcwd(),"logs",train_date)
    print("log_path is {}".format(log_path))

    #model path to save '.pth'file
    model_path = os.path.join(os.getcwd(),"models",train_date)
    print("model_path is {}".format(model_path))
    if os.path.isdir(model_path) == False:
            os.makedirs(model_path)

    #writer for tensorboard
    writer = SummaryWriter(log_path)
 
    # whether available GPU
    # device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    #device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    #define model (처음부터 학습시)
    print("처음부터 학습")
    model = MyModel()
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model.to(device)

    ####### 이전 학습한거에 추가적으로 학습시
    # print("이전에 하던거 부터 학습")
    # model_name = MyModel
    # # /content/models/230412_2145/0.pth
    # checkpoint_path = './12.pth'  #수정 필요
    # model = get_model(model_name,checkpoint_path)
    # model.to(device)

    #######

    #define loss funcion, optimizer, (scheduler)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(),lr=0.1, momentum=0.9,weight_decay=0.0001)
    #lr = 0.025, momentum = 0.9, weight_decay = 0.0005(original)
    # scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [81,122], gamma = 0.1)
    scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [81,122], gamma = 0.1)
    # 0.1 -> 0.01 (32K iterations = 81 epochs) -> 0.001 (48K iterations = 122 epochs)

    print("start training")
    #Train
    for epoch in range(epochs):

        model.train(True)
        len_trainbatch = len(train_dataloader)

        i = 0
        for inputs, targets in tqdm(train_dataloader):
            i = i+1
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            

            writer.add_scalar('train/loss', loss.item(), epoch*len_trainbatch+i) #iterations
        
        scheduler.step()    
        print("train loss of {}th epoch is {}".format(epoch,loss.item())) ## 수정 필요
 
        # Save every N(30) epoch
        save_epoch = 30
        if save_epoch > 0 and epoch % save_epoch == save_epoch-1:
            state_dict = model.state_dict()
            torch.save(state_dict, os.path.join(model_path, str(epoch+1)+'.pth')) #epoch+1로 바꾸기
        
    print('Finished Training')

    #save model after training
    torch.save(model.state_dict(), './model.pth') 

   

# Main

In [13]:

def main() :

    random.seed(53)
    #train(163, trainloader) #64k iterations = 163 epochs
    train(163, trainloader) #64k iterations = 163 epochs

   

if __name__ == '__main__':
    main()

log_path is /content/logs/230417_0053
model_path is /content/models/230417_0053
처음부터 학습
start training


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 0th epoch is 2.030055522918701


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 1th epoch is 1.5957404375076294


100%|██████████| 391/391 [00:37<00:00, 10.31it/s]


train loss of 2th epoch is 1.1459627151489258


100%|██████████| 391/391 [00:38<00:00, 10.09it/s]


train loss of 3th epoch is 0.8990737199783325


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 4th epoch is 0.8849180936813354


100%|██████████| 391/391 [00:38<00:00, 10.06it/s]


train loss of 5th epoch is 0.5908142924308777


100%|██████████| 391/391 [00:38<00:00, 10.06it/s]


train loss of 6th epoch is 0.9640769958496094


100%|██████████| 391/391 [00:38<00:00, 10.12it/s]


train loss of 7th epoch is 0.5910776853561401


100%|██████████| 391/391 [00:38<00:00, 10.14it/s]


train loss of 8th epoch is 0.7850844860076904


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 9th epoch is 0.4737507402896881


100%|██████████| 391/391 [00:38<00:00, 10.12it/s]


train loss of 10th epoch is 0.5467962026596069


100%|██████████| 391/391 [00:38<00:00, 10.11it/s]


train loss of 11th epoch is 0.6672261953353882


100%|██████████| 391/391 [00:38<00:00, 10.08it/s]


train loss of 12th epoch is 0.7558947205543518


100%|██████████| 391/391 [00:38<00:00, 10.08it/s]


train loss of 13th epoch is 0.47527456283569336


100%|██████████| 391/391 [00:38<00:00, 10.08it/s]


train loss of 14th epoch is 0.5281394124031067


100%|██████████| 391/391 [00:39<00:00,  9.92it/s]


train loss of 15th epoch is 0.28348448872566223


100%|██████████| 391/391 [00:38<00:00, 10.10it/s]


train loss of 16th epoch is 0.3247082829475403


100%|██████████| 391/391 [00:38<00:00, 10.09it/s]


train loss of 17th epoch is 0.39010632038116455


100%|██████████| 391/391 [00:38<00:00, 10.03it/s]


train loss of 18th epoch is 0.4111756682395935


100%|██████████| 391/391 [00:38<00:00, 10.06it/s]


train loss of 19th epoch is 0.40375709533691406


100%|██████████| 391/391 [00:38<00:00, 10.05it/s]


train loss of 20th epoch is 0.28552761673927307


100%|██████████| 391/391 [00:39<00:00, 10.02it/s]


train loss of 21th epoch is 0.29108160734176636


100%|██████████| 391/391 [00:39<00:00, 10.02it/s]


train loss of 22th epoch is 0.3915952444076538


100%|██████████| 391/391 [00:38<00:00, 10.06it/s]


train loss of 23th epoch is 0.41185373067855835


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 24th epoch is 0.2807082235813141


100%|██████████| 391/391 [00:39<00:00, 10.01it/s]


train loss of 25th epoch is 0.24417157471179962


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 26th epoch is 0.28556522727012634


100%|██████████| 391/391 [00:38<00:00, 10.22it/s]


train loss of 27th epoch is 0.4323420524597168


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 28th epoch is 0.2774924635887146


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 29th epoch is 0.3074567914009094


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 30th epoch is 0.28785640001296997


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 31th epoch is 0.29982703924179077


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 32th epoch is 0.3055756390094757


100%|██████████| 391/391 [00:38<00:00, 10.22it/s]


train loss of 33th epoch is 0.2561272382736206


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 34th epoch is 0.34997421503067017


100%|██████████| 391/391 [00:37<00:00, 10.29it/s]


train loss of 35th epoch is 0.3829968571662903


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 36th epoch is 0.25135284662246704


100%|██████████| 391/391 [00:38<00:00, 10.25it/s]


train loss of 37th epoch is 0.2361055612564087


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 38th epoch is 0.3471788167953491


100%|██████████| 391/391 [00:38<00:00, 10.25it/s]


train loss of 39th epoch is 0.2345007359981537


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 40th epoch is 0.21665093302726746


100%|██████████| 391/391 [00:38<00:00, 10.22it/s]


train loss of 41th epoch is 0.28711000084877014


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 42th epoch is 0.14265333116054535


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 43th epoch is 0.23265179991722107


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 44th epoch is 0.28829625248908997


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 45th epoch is 0.16264373064041138


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 46th epoch is 0.20407359302043915


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 47th epoch is 0.28001827001571655


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 48th epoch is 0.2927939295768738


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 49th epoch is 0.3890132009983063


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 50th epoch is 0.14497217535972595


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 51th epoch is 0.24497553706169128


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 52th epoch is 0.27573198080062866


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 53th epoch is 0.23567283153533936


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 54th epoch is 0.20213353633880615


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 55th epoch is 0.18982717394828796


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 56th epoch is 0.25167983770370483


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 57th epoch is 0.23272867500782013


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 58th epoch is 0.11121697723865509


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 59th epoch is 0.3144500255584717


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 60th epoch is 0.23309239745140076


100%|██████████| 391/391 [00:38<00:00, 10.22it/s]


train loss of 61th epoch is 0.21771900355815887


100%|██████████| 391/391 [00:39<00:00,  9.99it/s]


train loss of 62th epoch is 0.15888497233390808


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 63th epoch is 0.18713633716106415


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 64th epoch is 0.1998145878314972


100%|██████████| 391/391 [00:38<00:00, 10.27it/s]


train loss of 65th epoch is 0.2685139775276184


100%|██████████| 391/391 [00:38<00:00, 10.26it/s]


train loss of 66th epoch is 0.18115566670894623


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 67th epoch is 0.3548486828804016


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 68th epoch is 0.148898184299469


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 69th epoch is 0.25491124391555786


100%|██████████| 391/391 [00:38<00:00, 10.25it/s]


train loss of 70th epoch is 0.23495957255363464


100%|██████████| 391/391 [00:38<00:00, 10.25it/s]


train loss of 71th epoch is 0.24550509452819824


100%|██████████| 391/391 [00:38<00:00, 10.05it/s]


train loss of 72th epoch is 0.16642092168331146


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 73th epoch is 0.22447487711906433


100%|██████████| 391/391 [00:38<00:00, 10.27it/s]


train loss of 74th epoch is 0.20032057166099548


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 75th epoch is 0.14449357986450195


100%|██████████| 391/391 [00:38<00:00, 10.25it/s]


train loss of 76th epoch is 0.3646814227104187


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 77th epoch is 0.1536167562007904


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 78th epoch is 0.18261168897151947


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 79th epoch is 0.09992075711488724


100%|██████████| 391/391 [00:38<00:00, 10.22it/s]


train loss of 80th epoch is 0.3943110704421997


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 81th epoch is 0.09262105822563171


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 82th epoch is 0.19892096519470215


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 83th epoch is 0.04646584764122963


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 84th epoch is 0.08054593205451965


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 85th epoch is 0.03913252800703049


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 86th epoch is 0.06747995316982269


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 87th epoch is 0.034397948533296585


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 88th epoch is 0.05153390020132065


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 89th epoch is 0.06402061134576797


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 90th epoch is 0.035051118582487106


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 91th epoch is 0.06209911033511162


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 92th epoch is 0.09059469401836395


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 93th epoch is 0.017784034833312035


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 94th epoch is 0.027462180703878403


100%|██████████| 391/391 [00:38<00:00, 10.25it/s]


train loss of 95th epoch is 0.02615773119032383


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 96th epoch is 0.02271723374724388


100%|██████████| 391/391 [00:38<00:00, 10.28it/s]


train loss of 97th epoch is 0.017895257100462914


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 98th epoch is 0.009932123124599457


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 99th epoch is 0.018337052315473557


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 100th epoch is 0.007347973994910717


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 101th epoch is 0.013377875089645386


100%|██████████| 391/391 [00:38<00:00, 10.22it/s]


train loss of 102th epoch is 0.03766581788659096


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 103th epoch is 0.04017730802297592


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 104th epoch is 0.022167574614286423


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 105th epoch is 0.022118795663118362


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 106th epoch is 0.00732098612934351


100%|██████████| 391/391 [00:38<00:00, 10.16it/s]


train loss of 107th epoch is 0.036255400627851486


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 108th epoch is 0.009732259437441826


100%|██████████| 391/391 [00:38<00:00, 10.04it/s]


train loss of 109th epoch is 0.018545683473348618


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 110th epoch is 0.0278086569160223


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 111th epoch is 0.002464374527335167


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 112th epoch is 0.0069209253415465355


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 113th epoch is 0.0048287371173501015


100%|██████████| 391/391 [00:38<00:00, 10.11it/s]


train loss of 114th epoch is 0.022971123456954956


100%|██████████| 391/391 [00:38<00:00, 10.13it/s]


train loss of 115th epoch is 0.0344507060945034


100%|██████████| 391/391 [00:38<00:00, 10.16it/s]


train loss of 116th epoch is 0.004921280778944492


100%|██████████| 391/391 [00:38<00:00, 10.16it/s]


train loss of 117th epoch is 0.013057622127234936


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 118th epoch is 0.0033934537786990404


100%|██████████| 391/391 [00:39<00:00,  9.96it/s]


train loss of 119th epoch is 0.022056974470615387


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 120th epoch is 0.002788637299090624


100%|██████████| 391/391 [00:38<00:00, 10.11it/s]


train loss of 121th epoch is 0.023099152371287346


100%|██████████| 391/391 [00:38<00:00, 10.06it/s]


train loss of 122th epoch is 0.004274086561053991


100%|██████████| 391/391 [00:38<00:00, 10.14it/s]


train loss of 123th epoch is 0.002205303404480219


100%|██████████| 391/391 [00:38<00:00, 10.14it/s]


train loss of 124th epoch is 0.005525748245418072


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 125th epoch is 0.023769069463014603


100%|██████████| 391/391 [00:38<00:00, 10.13it/s]


train loss of 126th epoch is 0.026958713307976723


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 127th epoch is 0.015116879716515541


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 128th epoch is 0.010423840954899788


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 129th epoch is 0.0006666887202300131


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 130th epoch is 0.002875549253076315


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 131th epoch is 0.0017277036095038056


100%|██████████| 391/391 [00:38<00:00, 10.26it/s]


train loss of 132th epoch is 0.017254739999771118


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 133th epoch is 0.01914326101541519


100%|██████████| 391/391 [00:38<00:00, 10.20it/s]


train loss of 134th epoch is 0.009331694804131985


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 135th epoch is 0.007851017639040947


100%|██████████| 391/391 [00:38<00:00, 10.16it/s]


train loss of 136th epoch is 0.035486236214637756


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 137th epoch is 0.004480930045247078


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 138th epoch is 0.008675245568156242


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 139th epoch is 0.01963110640645027


100%|██████████| 391/391 [00:38<00:00, 10.16it/s]


train loss of 140th epoch is 0.005661597475409508


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 141th epoch is 0.000651311594992876


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 142th epoch is 0.0011121530551463366


100%|██████████| 391/391 [00:38<00:00, 10.13it/s]


train loss of 143th epoch is 0.0037680231034755707


100%|██████████| 391/391 [00:38<00:00, 10.18it/s]


train loss of 144th epoch is 0.012841668911278248


100%|██████████| 391/391 [00:38<00:00, 10.12it/s]


train loss of 145th epoch is 0.01655254140496254


100%|██████████| 391/391 [00:38<00:00, 10.13it/s]


train loss of 146th epoch is 0.00837524700909853


100%|██████████| 391/391 [00:38<00:00, 10.11it/s]


train loss of 147th epoch is 0.0033521712757647038


100%|██████████| 391/391 [00:38<00:00, 10.13it/s]


train loss of 148th epoch is 0.002184749348089099


100%|██████████| 391/391 [00:38<00:00, 10.17it/s]


train loss of 149th epoch is 0.01188563834875822


100%|██████████| 391/391 [00:38<00:00, 10.19it/s]


train loss of 150th epoch is 0.011057361960411072


100%|██████████| 391/391 [00:38<00:00, 10.15it/s]


train loss of 151th epoch is 0.003649247344583273


100%|██████████| 391/391 [00:38<00:00, 10.11it/s]


train loss of 152th epoch is 0.004863426089286804


100%|██████████| 391/391 [00:38<00:00, 10.14it/s]


train loss of 153th epoch is 0.01448771357536316


100%|██████████| 391/391 [00:38<00:00, 10.12it/s]


train loss of 154th epoch is 0.010781217366456985


100%|██████████| 391/391 [00:39<00:00,  9.96it/s]


train loss of 155th epoch is 0.015832137316465378


100%|██████████| 391/391 [00:38<00:00, 10.11it/s]


train loss of 156th epoch is 0.00821151863783598


100%|██████████| 391/391 [00:38<00:00, 10.08it/s]


train loss of 157th epoch is 0.01233386155217886


100%|██████████| 391/391 [00:38<00:00, 10.05it/s]


train loss of 158th epoch is 0.007265823893249035


100%|██████████| 391/391 [00:38<00:00, 10.09it/s]


train loss of 159th epoch is 0.0040213423781096935


100%|██████████| 391/391 [00:38<00:00, 10.13it/s]


train loss of 160th epoch is 0.006305375602096319


100%|██████████| 391/391 [00:38<00:00, 10.07it/s]


train loss of 161th epoch is 0.002765841316431761


100%|██████████| 391/391 [00:38<00:00, 10.05it/s]

train loss of 162th epoch is 0.017061877995729446
Finished Training


# Test

- model

100 100 Loss: 0.344 | Acc: 92.530% (9253/10000)


- 30

100 Loss: 0.549 | Acc: 82.450% (8245/10000)


- 60

100 100 Loss: 0.475 | Acc: 85.630% (8563/10000)

- 90

100 100 Loss: 0.285 | Acc: 92.240% (9224/10000)


- 120

100 100 Loss: 0.343 | Acc: 92.430% (9243/10000)

- 150

100 100 Loss: 0.341 | Acc: 92.460% (9246/10000)

In [24]:
def test():

    ####### not touch
    model_name = MyModel
    # /content/models/230412_2145/0.pth
    #/content/models/230412_2208/1.pth
    #/content/models/230416_0841/90.pth
    #/content/models/230417_0053/30.pth
    checkpoint_path = './models/230417_0053/150.pth'  #수정 필요
    mode = 'test' 
#     data_dir = "./test_data"
#     meta_path = "./answer.json"
    model = get_model(model_name,checkpoint_path)
    #######
    
#     # Create training and validation datasets
#     test_datasets = MyDataset(meta_path, data_dir, data_transforms['test'],is_train= False) #self,meta_path,root_dir,transform=None, is_train = True

#     # Create training and validation dataloaders
#     test_dataloader = torch.utils.data.DataLoader(test_datasets, batch_size=batch_size, shuffle=False, num_workers=4)

    test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
    
    testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)
    
    # Detect if we have a GPU available
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    # Send the model to GPU
    #multi-GPU
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model = model.to(device)

    # Set model as evaluation mode
    for param in model.parameters():
        param.requires_grad = False
    model.eval()
    
    # Inference
    test_loss = 0
    correct = 0
    total = 0
    criterion = nn.CrossEntropyLoss()
    with torch.no_grad():
        batch_idx = 0
        for inputs, targets in tqdm(testloader):
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            
            test_loss += loss.item()
            _, predicted = outputs.max(1)
            total += targets.size(0)
            correct += predicted.eq(targets).sum().item()
            batch_idx = batch_idx + 1
            print(batch_idx, len(testloader), 'Loss: %.3f | Acc: %.3f%% (%d/%d)'
                         % (test_loss/(batch_idx+1), 100.*correct/total, correct, total))



In [25]:

def main() :

    random.seed(53)
    test()

   
if __name__ == '__main__':
    main()

start loading model parameter
model parameter loaded
Files already downloaded and verified


  2%|▏         | 2/100 [00:00<00:17,  5.47it/s]

1 100 Loss: 0.119 | Acc: 95.000% (95/100)
2 100 Loss: 0.205 | Acc: 93.500% (187/200)


  6%|▌         | 6/100 [00:00<00:07, 12.48it/s]

3 100 Loss: 0.181 | Acc: 94.667% (284/300)
4 100 Loss: 0.230 | Acc: 93.750% (375/400)
5 100 Loss: 0.257 | Acc: 93.800% (469/500)
6 100 Loss: 0.247 | Acc: 94.167% (565/600)


 11%|█         | 11/100 [00:00<00:04, 18.42it/s]

7 100 Loss: 0.267 | Acc: 93.857% (657/700)
8 100 Loss: 0.283 | Acc: 93.375% (747/800)
9 100 Loss: 0.301 | Acc: 93.222% (839/900)
10 100 Loss: 0.312 | Acc: 93.100% (931/1000)
11 100 Loss: 0.301 | Acc: 93.273% (1026/1100)
12

 17%|█▋        | 17/100 [00:01<00:03, 22.54it/s]

 100 Loss: 0.304 | Acc: 93.167% (1118/1200)
13 100 Loss: 0.307 | Acc: 93.000% (1209/1300)
14 100 Loss: 0.306 | Acc: 93.000% (1302/1400)
15 100 Loss: 0.303 | Acc: 93.000% (1395/1500)
16 100 Loss: 0.296 | Acc: 93.125% (1490/1600)
17 100 Loss: 0.303 | Acc: 93.176% (1584/1700)


 23%|██▎       | 23/100 [00:01<00:03, 24.30it/s]

18 100 Loss: 0.309 | Acc: 92.889% (1672/1800)
19 100 Loss: 0.315 | Acc: 92.684% (1761/1900)
20 100 Loss: 0.335 | Acc: 92.600% (1852/2000)
21 100 Loss: 0.340 | Acc: 92.476% (1942/2100)
22 100 Loss: 0.346 | Acc: 92.227% (2029/2200)
23 100 Loss: 0.346 | Acc: 92.174% (2120/2300)


 29%|██▉       | 29/100 [00:01<00:02, 25.19it/s]

24 100 Loss: 0.351 | Acc: 92.083% (2210/2400)
25 100 Loss: 0.358 | Acc: 92.000% (2300/2500)
26 100 Loss: 0.371 | Acc: 91.769% (2386/2600)
27 100 Loss: 0.364 | Acc: 91.852% (2480/2700)
28 100 Loss: 0.364 | Acc: 91.929% (2574/2800)
29 100 Loss: 0.363 | Acc: 91.966% (2667/2900)


 35%|███▌      | 35/100 [00:01<00:02, 26.01it/s]

30 100 Loss: 0.361 | Acc: 91.967% (2759/3000)
31 100 Loss: 0.356 | Acc: 91.968% (2851/3100)
32 100 Loss: 0.353 | Acc: 92.031% (2945/3200)
33 100 Loss: 0.353 | Acc: 92.061% (3038/3300)
34 100 Loss: 0.349 | Acc: 92.176% (3134/3400)
35 100 Loss: 0.351 | Acc: 92.114% (3224/3500)


 39%|███▉      | 39/100 [00:01<00:02, 27.61it/s]

36 100 Loss: 0.352 | Acc: 92.111% (3316/3600)
37 100 Loss: 0.359 | Acc: 92.000% (3404/3700)
38 100 Loss: 0.363 | Acc: 91.921% (3493/3800)
39 100 Loss: 0.360 | Acc: 92.026% (3589/3900)
40 100 Loss: 0.361 | Acc: 92.025% (3681/4000)
41 100 Loss: 0.366 | Acc: 91.951% (3770/4100)
42 100 Loss: 0.366 | Acc: 91.976% (3863/4200)


 46%|████▌     | 46/100 [00:02<00:01, 28.62it/s]

43 100 Loss: 0.362 | Acc: 92.070% (3959/4300)
44 100 Loss: 0.362 | Acc: 92.159% (4055/4400)
45 100 Loss: 0.359 | Acc: 92.200% (4149/4500)
46 100 Loss: 0.362 | Acc: 92.109% (4237/4600)
47 100 Loss: 0.361 | Acc: 92.128% (4330/4700)
48 100 Loss: 0.366 | Acc: 92.000% (4416/4800)


 52%|█████▏    | 52/100 [00:02<00:01, 26.58it/s]

49 100 Loss: 0.360 | Acc: 92.082% (4512/4900)
50 100 Loss: 0.363 | Acc: 92.060% (4603/5000)
51 100 Loss: 0.359 | Acc: 92.137% (4699/5100)
52 100 Loss: 0.359 | Acc: 92.173% (4793/5200)
53 100 Loss: 0.356 | Acc: 92.226% (4888/5300)
54 100 Loss: 0.355 | Acc: 92.241% (4981/5400)


 59%|█████▉    | 59/100 [00:02<00:01, 27.16it/s]

55 100 Loss: 0.358 | Acc: 92.218% (5072/5500)
56 100 Loss: 0.357 | Acc: 92.268% (5167/5600)
57 100 Loss: 0.358 | Acc: 92.228% (5257/5700)
58 100 Loss: 0.353 | Acc: 92.293% (5353/5800)
59 100 Loss: 0.356 | Acc: 92.237% (5442/5900)
60 100 Loss: 0.355 | Acc: 92.217% (5533/6000)


 66%|██████▌   | 66/100 [00:02<00:01, 27.96it/s]

61 100 Loss: 0.355 | Acc: 92.213% (5625/6100)
62 100 Loss: 0.353 | Acc: 92.242% (5719/6200)
63 100 Loss: 0.351 | Acc: 92.270% (5813/6300)
64 100 Loss: 0.348 | Acc: 92.312% (5908/6400)
65 100 Loss: 0.345 | Acc: 92.323% (6001/6500)
66 100 Loss: 0.342 | Acc: 92.379% (6097/6600)


 72%|███████▏  | 72/100 [00:03<00:01, 26.88it/s]

67 100 Loss: 0.340 | Acc: 92.388% (6190/6700)
68 100 Loss: 0.343 | Acc: 92.368% (6281/6800)
69 100 Loss: 0.343 | Acc: 92.362% (6373/6900)
70 100 Loss: 0.343 | Acc: 92.343% (6464/7000)
71 100 Loss: 0.344 | Acc: 92.338% (6556/7100)
72 100 Loss: 0.343 | Acc: 92.347% (6649/7200)


 78%|███████▊  | 78/100 [00:03<00:00, 27.26it/s]

73 100 Loss: 0.342 | Acc: 92.342% (6741/7300)
74 100 Loss: 0.342 | Acc: 92.365% (6835/7400)
75 100 Loss: 0.345 | Acc: 92.373% (6928/7500)
76 100 Loss: 0.346 | Acc: 92.368% (7020/7600)
77 100 Loss: 0.346 | Acc: 92.390% (7114/7700)
78 100 Loss: 0.347 | Acc: 92.397% (7207/7800)


 81%|████████  | 81/100 [00:03<00:00, 26.99it/s]

79 100 Loss: 0.344 | Acc: 92.418% (7301/7900)
80 100 Loss: 0.344 | Acc: 92.412% (7393/8000)
81 100 Loss: 0.343 | Acc: 92.432% (7487/8100)
82 100 Loss: 0.340 | Acc: 92.463% (7582/8200)
83 100 Loss: 0.342 | Acc: 92.434% (7672/8300)
84 100 Loss: 0.341 | Acc: 92.405% (7762/8400)


 88%|████████▊ | 88/100 [00:03<00:00, 27.16it/s]

85 100 Loss: 0.341 | Acc: 92.376% (7852/8500)
86 100 Loss: 0.344 | Acc: 92.349% (7942/8600)
87 100 Loss: 0.343 | Acc: 92.356% (8035/8700)
88 100 Loss: 0.342 | Acc: 92.398% (8131/8800)
89 100 Loss: 0.341 | Acc: 92.382% (8222/8900)
90 100 Loss: 0.340 | Acc: 92.422% (8318/9000)


 96%|█████████▌| 96/100 [00:03<00:00, 28.65it/s]

91 100 Loss: 0.339 | Acc: 92.451% (8413/9100)
92 100 Loss: 0.338 | Acc: 92.467% (8507/9200)
93 100 Loss: 0.341 | Acc: 92.398% (8593/9300)
94 100 Loss: 0.342 | Acc: 92.372% (8683/9400)
95 100 Loss: 0.341 | Acc: 92.389% (8777/9500)
96 100 Loss: 0.342 | Acc: 92.375% (8868/9600)


100%|██████████| 100/100 [00:04<00:00, 24.30it/s]

97 100 Loss: 0.340 | Acc: 92.423% (8965/9700)
98 100 Loss: 0.340 | Acc: 92.439% (9059/9800)
99 100 Loss: 0.340 | Acc: 92.444% (9152/9900)
100 100 Loss: 0.341 | Acc: 92.460% (9246/10000)
